File: packages/device/src/device-info.ts
import { execSync } from 'node:child_process';
import * as os from 'node:os';
import * as ort from 'onnxruntime-node';
import { getGpuInfo } from './gpu/gpu.ts';
import type { GpuInfo } from './gpu/types.ts';
import { fmtBytes } from './utils.ts';

export interface OrtBackend {
	name: string;
	bundled: boolean;
}

export interface DeviceInfo {
	platform: {
		os: string;
		arch: string;
		release: string;
		hostname: string;
		runtime: string;
		runtimeVersion: string;
	};
	cpu: {
		model: string;
		cores: number;
		speedMHz: number;
	};
	memory: {
		total: string;
		free: string;
		processHeapUsed: string;
	};
	gpu: GpuInfo[];
	ort: {
		version: string;
		backends: OrtBackend[];
	};
}

export async function getDeviceInfo(): Promise<DeviceInfo> {
	const cpus = os.cpus();
	const backends: OrtBackend[] = (ort.listSupportedBackends?.() ?? []).map(
		(b) => ({
			name: b.name,
			bundled: b.bundled,
		}),
	);

	return {
		platform: {
			os: process.platform,
			arch: process.arch,
			release: os.release(),
			hostname: os.hostname(),
			runtime: 'bun',
			runtimeVersion: process.version,
		},
		cpu: {
			model: cpus[0]?.model ?? 'unknown',
			cores: cpus.length,
			speedMHz: cpus[0]?.speed ?? 0,
		},
		memory: {
			total: fmtBytes(os.totalmem()),
			free: fmtBytes(os.freemem()),
			processHeapUsed: fmtBytes(process.memoryUsage().heapUsed),
		},
		gpu: getGpuInfo(),
		ort: {
			version: ort.env.versions?.common ?? 'unknown',
			backends,
		},
	};
}

File: packages/device-rs/src/device_info.rs
use serde::{Deserialize, Serialize};
use specta::Type;

use crate::gpu::GpuInfo;

#[derive(Debug, Clone, Serialize, Deserialize, Type)]
#[serde(rename_all = "camelCase")]
pub struct DeviceInfo {
    pub platform: PlatformInfo,
    pub cpu: CpuInfo,
    pub memory: MemoryInfo,
    pub gpu: Vec<GpuInfo>,
    pub ort: OrtInfo,
}

#[derive(Debug, Clone, Serialize, Deserialize, Type)]
#[serde(rename_all = "camelCase")]
pub struct PlatformInfo {
    pub os: String,
    pub arch: String,
    pub release: String,
    pub hostname: String,
    pub runtime: String,
    pub runtime_version: String,
    #[serde(skip_serializing_if = "Option::is_none")] // | undefined
    pub node_version: Option<String>, // T | None(null)
}

#[derive(Debug, Clone, Serialize, Deserialize, Type)]
#[serde(rename_all = "camelCase")]
pub struct CpuInfo {
    pub model: String,
    pub cores: u32,
    #[serde(rename = "speedMHz")]
    pub speed_mhz: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize, Type)]
#[serde(rename_all = "camelCase")]
pub struct MemoryInfo {
    pub total: String,
    pub free: String,
    pub process_heap_used: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, Type)]
#[serde(rename_all = "camelCase")]
pub struct OrtInfo {
    pub version: String,
    pub backends: Vec<OrtBackend>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Type)]
#[serde(rename_all = "camelCase")]
pub struct OrtBackend {
    pub name: String,
    pub bundled: bool,
}
